implementation/core/src/main/java/io/smallrye/faulttolerance/core/circuit/breaker/CompletionStageCircuitBreaker.java

```diff
@@ -124,24 +116,36 @@ private CompletionException onFailure(State state, Throwable e) {
 
     private CompletionStage<V> inHalfOpen(InvocationContext<CompletionStage<V>> target, CircuitBreaker.State state) {
         try {
-            CompletionStage<V> result = delegate.apply(target);
-            metricsRecorder.circuitBreakerSucceeded();
+            CompletableFuture<V> result = new CompletableFuture<>();
 
-            int successes = state.consecutiveSuccesses.incrementAndGet();
-            if (successes >= successThreshold) {
-                long now = System.nanoTime();
-                closedStart = now;
-                previousHalfOpenTime.addAndGet(now - halfOpenStart);
+            delegate.apply(target).whenComplete((value, error) -> {
+                if (error != null) {
+                    metricsRecorder.circuitBreakerFailed();
```

Comment:
Maybe worth pulling out the three lines here to a separate method?

Reply:
I'd prefer not. I tend to not create methods that can possibly only be called from a single place.

Reply:
you have exactly the same lines in 145-147

Reply:
Ah, true. That's in fact the same for all `CompletionStage*` strategies, and something I'd like to address in another way (later): the `Invocation` strategy would be, in the async flow, wrapped into an exception-handling strategy, which would catch exceptions thrown by the `Invocation` and propagate them to the `CompletableFuture`. No other `CompletionStage*` strategy would have to do that, anymore.

Reply:
Filed https://github.com/smallrye/smallrye-fault-tolerance/issues/202

Reply:
:+1: